frontend/simpleRedux.js

```diff
@@ -5,6 +5,89 @@ const {CARDANOLITE_CONFIG} = require('./frontendConfigLoader')
 let state, middlewares
 const root = {}
 
+const reconcile = (existingNode, virtualNode) => {
+  if (existingNode.tagName !== virtualNode.tagName) return virtualNode
+
+  const existingChildren = existingNode.children
+    ? Array.from(existingNode.children).filter((n) => n instanceof HTMLElement)
+    : []
+  const virtualChildren = virtualNode.children
+    ? Array.from(virtualNode.children).filter((n) => n instanceof HTMLElement)
+    : []
+
+  // attributes is a NamedNodeMap, which does not support same api as an array
+  for (let i = 0; i < virtualNode.attributes.length; i++) {
+    existingNode.setAttribute(virtualNode.attributes[i].name, virtualNode.attributes[i].value)
+  }
+  // remove attributes no longer valid
+  for (let i = 0; i < existingNode.attributes.length; i++) {
+    if (!virtualNode.hasAttribute(existingNode.attributes[i].name)) {
+      existingNode.removeAttribute(existingNode.attributes[i].name)
+    }
+  }
+  // without children, set inner content and be done
+  if (!virtualChildren.length) {
+    existingNode.innerHTML = virtualNode.innerHTML
+    return existingNode
+  }
+
+  // recursively reconcile children
+  const existingKeyedChildren = new Map()
+  existingChildren.forEach(
+    (child) => child.dataset.key && existingKeyedChildren.set(child.dataset.key, child)
+  )
+  const virtualKeyedChildren = new Map()
+  virtualChildren.forEach(
+    (child) => child.dataset.key && virtualKeyedChildren.set(child.dataset.key, child)
+  )
+
+  while (existingChildren.length && virtualChildren.length) {
+    const currentChild = existingChildren[0]
+
+    // Keyed children that don't exist in new virtual DOM are removed. The rest are skipped until
+    // encountered in the virtual DOM, unless the key matches the one on the 'nextChild' virtual
+    // node - this minimizes the amount of reflows, as it allows us to keep the original node,
+    // and ensures we do not lose focus on keyed inputs which did not move in the DOM.
+    if (currentChild.dataset.key && virtualChildren[0].dataset.key !== currentChild.dataset.key) {
+      !virtualKeyedChildren.has(currentChild.dataset.key) && existingNode.removeChild(currentChild)
+      existingChildren.shift()
+      continue
+    }
+
+    const nextChild = virtualChildren.shift()
```

Comment:
What is the cost for shift() - is it O(1) or O(n) - in the latter case, the whole reconciliation takes O(n^2) #sad.

Reply:
I got curious, so I found out:

```
const a = []

for (let i = 0; i < 1000000; i++) {
  a.push(i)
}

while (a.length > 0) {
  // if I do this, it takes 70ms
  //a.pop()
  // if I do this, it takes ages
  //a.shift()
}
```